### PR TITLE
Feat/add uhpo crate rework

### PR DESCRIPTION
**Pull request by jayendramadaram**
This Pr is First pr in series of broken Pr's from #9 

### Changes Made
- Added `payout-update` Transaction builder.
- Offers way to Schnorr Sign `payout-update`  Transactions.
- removed dependency from nigiri from testcase perspective.
- includes unit tests for `payout-update`.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,4 +18,5 @@ proposal/proposal.blg
 simulations/logs/*
 simulations/*.toolbox
 venv
-target
\ No newline at end of file
+target
+todo
```
Comment (pool2win) [inline .gitignore]: Move this to .git/config/ignore if this is your personal ignore file. Other contributors don't have this file.

File: node/connection/src/lib.rs [hunk 2]
```diff
@@ -343,10 +344,7 @@ mod tests {
     use tokio_util::bytes::Bytes;
 
     #[tokio::test]
-    // TODO: This is a smoke test and we need to improve how we wait for listen to complete before we invoke connect.
     async fn it_should_run_connect_without_errors() {
-        use tokio::time::{sleep, Duration};
```
Comment (pool2win) [inline node/connection/src/lib.rs]: Why did you remove these? This change doesn't seem related to the UHPO work, so if we need this change to go in, let's make it a separate PR.

File: node/connection/src/lib.rs [hunk 3]
```diff
@@ -358,29 +356,34 @@ mod tests {
...
                 send_to_all_tx,
-                Arc::new(Notify::new()),
+                notify_listen_cloned,
             )
             .await;
         });
 
-        // TODO: Fix this smoke test! Kill the sleep in this smoke test.
-        sleep(Duration::from_millis(100)).await;
+        notify.notified().await;
```
Comment (pool2win) [inline node/connection/src/lib.rs]: I think this change of notifier is valuable, but can you please move this into a separate PR so we can test and review it as a unit.

File: node/uhpo/docs/spec.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
+# Braidpool UHPO Spec
```
Comment (pool2win) [inline node/uhpo/docs/spec.md]: We should not have this design doc in the repo. I am not sure the design will always remain in sync with the code, and thus will cause confusion. Let's move this into a gist and link the gist from the PR description. This way the history of this discussion is maintained and we are not burdened to keep it in sync with the code as it evolves.

File: node/uhpo/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            }
+            UhpoError::Other(msg) => write!(f, "Other error: {}", msg),
+        }
+    }
+}
+
+impl std::error::Error for UhpoError {}
+
+impl UhpoError {
+    pub fn new_other_error(message: &str) -> Self {
```
Comment (pool2win) [inline node/uhpo/src/error.rs]: Why not simply call this method `new`?

**Review by pool2win — CHANGES_REQUESTED** [review]
@jayendramadaram - I left some initial comments that need to be addressed. I am still going through payout_update.rs and will follow up with another review.

**Comment by pool2win** [comment]
I see what happened with the changes for notifier. It seems like there is a problem with this PR where the commits from stutxo have crept into this PR. We need to clear this up, so that only your commits relevant to this PR are in this PR.

File: node/uhpo/src/payout_update.rs [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+pub struct PayoutUpdate<'a> {
+    transaction: Transaction,
+    coinbase_txout: &'a TxOut,
+    prev_update_txout: Option<&'a TxOut>,
+}
+
+impl<'a> PayoutUpdate<'a> {
+    pub fn new(
+        prev_update_tx: Option<&'a Transaction>,
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: I think we can let these arguments can be passed in without using references. I think the client can be required to transfer ownership to the payout update value. If the client wants to retain ownership, then they can Clone the value and pass it in here.

The reason to use ownership transfer instead of references is to keep the code simpler and therefore easier to maintain and use in the future.

Let's try to do this, and I can review again.

**Review by pool2win — CHANGES_REQUESTED** [review]
My second review here, about using ownership transfer instead of references with lifetimes.

File: node/uhpo/src/payout_update.rs (hunk 6 above)
Comment (jayendramadaram) [reply node/uhpo/src/payout_update.rs]: I've implemented the changes as suggested,
- Removed lifetime parameters from the PayoutUpdate struct.
- Modified the PayoutUpdate struct to take ownership of Transaction objects
- Adjusted internal methods to work with owned TxOut instances

```rust
pub struct PayoutUpdate {
    transaction: Transaction,
    coinbase_txout: TxOut,
    prev_update_txout: Option<TxOut>,
}

impl PayoutUpdate {
    pub fn new(
        prev_update_tx: Option<Transaction>,
        coinbase_tx: Transaction,
        next_out_address: Address,
        projected_fee: Amount,
    ) -> Result<Self, Error> {
        let prev_update_txout = prev_update_tx.as_ref().map(|tx| tx.output[0].clone());
        let coinbase_txout = coinbase_tx.output[0].clone();

        // ... rest of the implementation ...
    }

    // ... other methods updated accordingly ...
}
```

**Comment by jayendramadaram** [comment]
I have, made above suggested changes, latest commit includes

- removal of personal ignore files from .gitignore
- removal of spec docs
- This PR is rebased to remove commits from stutxo
- Modified the PayoutUpdate struct to take ownership of Transaction objects

File: .gitignore [hunk 7]
```diff
@@ -15,5 +15,4 @@ proposal/proposal.bbl
 proposal/proposal.blg
 simulations/logs/*
 simulations/*.toolbox
-venv
-target
\ No newline at end of file
+venv
```
Comment (pool2win) [inline .gitignore]: The venv should be in your private .git/info/exclude

File: node/uhpo/src/payout_update.rs [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    }
+}
+
+fn build_transaction(
+    coinbase_tx: &Transaction,
+    prev_update_tx: Option<&Transaction>,
+    next_out_address: Address,
+    projected_fee: Amount,
+    coinbase_txout: &TxOut,
+    prev_update_txout: Option<&TxOut>,
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: We don't need any of these arguments to be a reference.

Here's a working patch that shows how. https://github.com/pool2win/braidpool/commit/638d3c74d6e6695b535f241985ce3de010ed756b

File: node/uhpo/src/payout_update.rs [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            witness: Witness::default(),
+        });
+    }
+
+    payout_update_tx.output.push(TxOut {
+        value: total_amount - projected_fee,
+        script_pubkey: next_out_address.script_pubkey(),
+    });
+
+    Ok(payout_update_tx)
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: There are a lot of individual steps we are taking here. Maybe we should use a builder pattern here again and each of the separate blocks can be turned into separate functions? That will also result in the `build_transaction` function either disappearing or getting simpler.

https://rust-unofficial.github.io/patterns/patterns/creational/builder.html

File: node/uhpo/src/payout_update.rs [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+fn add_signature(
+    transaction: &mut Transaction,
+    input_idx: usize,
+    prevouts: &[&TxOut],
+    private_key: &SecretKey,
+    tweak: &Option<Scalar>,
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: Again, references. My understanding right now with Rust is - use move, it is efficient in Rust, use refs only if really needed.

File: node/uhpo/src/payout_update.rs [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+) -> Result<(), UhpoError> {
+    let secp = Secp256k1::new();
+
+    let keypair: Keypair = match tweak {
+        Some(tweak) => private_key.keypair(&secp).add_xonly_tweak(&secp, tweak)?,
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: We should differentiate the error here and the one on line 158, so that the client knows where the signing failed.

File: node/uhpo/src/payout_update.rs [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+        payout_update
+            .add_prev_update_sig(&keypair.secret_key(), &None)
+            .expect("Failed to add prev update signature");
+
+        assert_eq!(payout_update.transaction.input[0].witness.len(), 1);
+        assert_eq!(payout_update.transaction.input[1].witness.len(), 1);
+    }
+
+    fn create_dummy_transaction() -> Transaction {
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: Move this function to the top, below the setup. All support functions should come before the test functions.

File: node/uhpo/src/payout_update.rs [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        assert_eq!(payout_update.transaction.output.len(), 1);
+        assert_eq!(
+            payout_update.transaction.output[0].value,
+            Amount::from_sat(50000000 * 2) - projected_fee
+        );
+        assert!(payout_update.prev_update_txout.is_some());
+    }
+
+    #[test]
+    fn test_add_signature() {
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: We should add tests for the code paths where:
1. No tweak is passed
2. There is an error on line 140
3. There is an error on line 158
4. Add test where signature verification fails - you might need to use mocks for this

File: node/uhpo/src/payout_update.rs [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+
+        add_signature(&mut self.transaction, 1, &prevouts, private_key, tweak)
+    }
+
+    pub fn build(self) -> Transaction {
+        self.transaction
+    }
+}
+
+fn build_transaction(
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: Add doc comment

File: node/uhpo/src/payout_update.rs [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+    ) -> Result<(), UhpoError> {
+        let prevouts = match &self.prev_update_txout {
+            Some(prev_update_txout) => vec![&self.coinbase_txout, prev_update_txout],
+            None => vec![&self.coinbase_txout],
+        };
+
+        add_signature(&mut self.transaction, 0, &prevouts, private_key, tweak)
+    }
+
+    pub fn add_prev_update_sig(
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: Add doc comment

File: node/uhpo/src/payout_update.rs [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+        )?;
+
+        Ok(PayoutUpdate {
+            transaction: payout_update_tx,
+            coinbase_txout,
+            prev_update_txout,
+        })
+    }
+
+    pub fn add_coinbase_sig(
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs]: Add doc comment

File: node/uhpo/src/payout_update.rs [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+    secp256k1::{Message, Scalar, SecretKey},
+    sighash::{Prevouts, SighashCache},
+    transaction::Version,
+    Address, Amount, OutPoint, ScriptBuf, Sequence, TapSighashType, Transaction, TxIn, TxOut,
+    Witness,
+};
+
+use crate::error::UhpoError;
+
+pub struct PayoutUpdate {
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs:11]: Add doc comment

**Review by pool2win — CHANGES_REQUESTED** [review]
Nice progress. A few more things we need to do. See individual comments.

File: node/uhpo/src/payout_update.rs (hunk 9 above)
Comment (jayendramadaram) [reply node/uhpo/src/payout_update.rs]: I 'have Implemented changes as suggested,
- added new `TransactionBuilder` which can be used but both `PayoutSettlement` and `PayoutUpdate`

File: node/uhpo/src/payout_update.rs (hunk 10 above)
Comment (jayendramadaram) [reply node/uhpo/src/payout_update.rs]: I have changes private_key from not to take as reference, but other two prevouts and tweak are meant to be passed as reference to rust-bitcoin's `SighashCache.taproot_key_spend_signature_hash` and `Keypair.add_xonly_tweak`, hence using  continuing with reference only for these two params.

File: node/uhpo/src/payout_update.rs (hunk 11 above)
Comment (jayendramadaram) [reply node/uhpo/src/payout_update.rs]: Added new errors as suggested

- UhpoError::KeypairCreationError
- UhpoError::SignatureVerificationError

File: node/uhpo/src/payout_update.rs (hunk 13 above)
Comment (jayendramadaram) [reply node/uhpo/src/payout_update.rs]: With latest changes made,
- new testcases are added to cover tweak keypair case.
- mock `KeypairCreationError` and `SignatureVerificationError` using `mockall` crate and making SecretKey , Secp256k1 and Keypair Generic.

**Comment by jayendramadaram** [comment]
I have implemented the suggested changes in the latest commit. The updates include:

1. Added  doc strings and comments to code .
2. Implemented a separate `TransactionBuilder` for constructing Bitcoin transactions in `PayoutUpdate`, improving modularity and reusability.
3. Moved `add_signatures` function to `src/crypto/signature.rs`, as it's used by both `PayoutUpdate` and `PayoutSettlement`.
4. Introduced new, more descriptive error types.
5. Implemented mock tests to simulate `rust-bitcoin` instances returning errors.

File: node/uhpo/src/crypto/signature.rs [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+            input: vec![],
+            output: vec![TxOut {
+                value: Amount::from_sat(50000000),
+                script_pubkey: ScriptBuf::new(),
+            }],
+        }
+    }
+
+    #[test]
+    fn test_add_signatures_key_creation_fails() {
```
Comment (pool2win) [inline node/uhpo/src/crypto/signature.rs]: Let's rename test to 

`test_add_signature_with_invalid_tweak_error_signature_creation_should_fail`

basically

`test_<fn>_<under_which_conditions>_<what_should_happen>`

File: node/uhpo/src/crypto/signature.rs [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+            mock_secret_key,
+            tweak.as_ref(),
+            &mock_secp,
+        );
+
+        assert!(matches!(result, Err(UhpoError::KeypairCreationError(_))));
+    }
+
+    #[test]
+    fn test_add_signatures_signature_verification_fails() {
```
Comment (pool2win) [inline node/uhpo/src/crypto/signature.rs]: fn name here again,

`test_add_signature_with_a_bad_signature_should_fail_during_verification`

File: node/uhpo/src/crypto/signature.rs [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+    // Add the signature to the transaction input
+    let mut vec_sig = signature.serialize().to_vec();
+    vec_sig.push(0x01);
+    transaction.input[input_idx].witness.push(vec_sig);
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod mock_tests {
```
Comment (pool2win) [inline node/uhpo/src/crypto/signature.rs:148]: Let's add a success case as well. So that we have a unit test for the good case.

File: node/uhpo/src/payout_update.rs [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    pub fn build(self) -> Transaction {
+        self.transaction
+    }
+}
+
+// unit tests
+#[cfg(test)]
+mod tests {
```
Comment (pool2win) [inline node/uhpo/src/payout_update.rs:145]: Let's add a test to see we handle errors properly from `add_signature`. If there is a match on the error, we should have one test case for each.

**Review by pool2win — CHANGES_REQUESTED** [review]
Nice! I am still not sure if the new approach adds a function call overhead - I'll look at that again.

For now a few things:

1. CI is failing.
2. We should add the failure/success test cases appropriately. I left comments in the PR review for where we need more tests.

**Comment by jayendramadaram** [comment]
Fixes made from latest commit,
- changed testcase names as requested
- added a mock error testcase which mocks `KeyCreationError(_)` on `add_coinbase_sig` and `add_prev_update_sig` methods on `PayoutUpdate`. I feel single mock testcase should be fine, ignoring testcase for `SignatureVerificationError`.
